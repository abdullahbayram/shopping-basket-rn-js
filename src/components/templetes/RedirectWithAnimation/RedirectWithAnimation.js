import React, { useCallback, useEffect, useRef, useState } from 'react';
import { StyleSheet, View, Animated, Text, BackHandler } from 'react-native';
import { useNavigation } from '@react-navigation/native';
import PropTypes from 'prop-types';
import Button from '../../atoms/Button/Button';

const RedirectWithAnimation = ({ message, duration = 5000 }) => {
  const navigation = useNavigation();
  const [seconds, setSeconds] = useState(duration / 1000); // Countdown in seconds
  const progress = useRef(new Animated.Value(0)).current; // Animation progress
  const redirectTo = 'ProductList';

  const resetNavigation = useCallback(() => {
    navigation.reset({
      index: 0,
      routes: [{ name: redirectTo }],
    });
  }, [navigation]);

  useEffect(() => {
    // Disable Android back button
    const backHandler = BackHandler.addEventListener('hardwareBackPress', () => true);

    Animated.timing(progress, {
      toValue: 1, // Animate to 100% progress
      duration,
      useNativeDriver: false, // Required for width animations
    }).start();

    // Countdown logic
    const interval = setInterval(() => {
      setSeconds((prev) => {
        if (prev === 1) {
          clearInterval(interval);
        }
        return prev - 1;
      });
    }, 1000);
    // Redirect logic
    const timer = setTimeout(() => {
      resetNavigation();
    }, duration);

    return () => {
      backHandler.remove();
      clearInterval(interval);
      clearTimeout(timer);
    };
  }, [navigation, progress, duration, redirectTo, resetNavigation]);

  const widthInterpolated = progress.interpolate({
    inputRange: [0, 1],
    outputRange: ['0%', '100%'],
  });

  return (
    <View style={styles.container}>
      <View style={styles.textContainer}>
        <Text style={styles.title}>{message}</Text>
        <Text style={styles.message}>
          Redirecting to product list in {seconds} second{seconds > 1 ? 's' : ''}...
        </Text>
      </View>
      <View style={styles.progressBarContainer}>
        <Animated.View style={[styles.progressBar, { width: widthInterpolated }]} />
      </View>
      <Button mode="contained" onPress={resetNavigation}>
        Go to Products
      </Button>
    </View>
  );
};

RedirectWithAnimation.propTypes = {
  message: PropTypes.string.isRequired,
  duration: PropTypes.number,
};

export default RedirectWithAnimation;

const styles = StyleSheet.create({
  container: {
    flex: 1,
    justifyContent: 'center',
    alignItems: 'center',
    backgroundColor: '#fff',
  },
  textContainer: {
    marginBottom: 20,
    alignItems: 'center',
  },
  title: {
    fontWeight: 'bold',
    fontSize: 24,
    marginBottom: 8,
    textAlign: 'center',
  },
  message: {
    fontSize: 16,
    color: '#666',
  },
  progressBarContainer: {
    width: '80%',
    height: 10,
    backgroundColor: '#e0e0e0',
    borderRadius: 5,
    overflow: 'hidden', // Ensure the progress stays inside the container
    marginBottom: 16,
  },
  progressBar: {
    height: '100%',
    backgroundColor: '#76c7c0',
    borderRadius: 5,
  },
});
