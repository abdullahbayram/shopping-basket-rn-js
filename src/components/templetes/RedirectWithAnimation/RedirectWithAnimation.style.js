import { StyleSheet } from 'react-native';
import { typography } from '../../../constants/theme';

const createStyles = (colors, spacing) =>
  StyleSheet.create({
    container: {
      flex: 1,
      justifyContent: 'center',
      alignItems: 'center',
      backgroundColor: colors.background,
    },
    title: {
      fontWeight: typography.bold,
      fontSize: typography.xl,
      marginBottom: spacing.special,
      textAlign: 'center',
      color: colors.textPrimary,
    },
    textContainer: {
      marginBottom: spacing.lg,
      alignItems: 'center',
    },
    message: {
      fontSize: typography.default,
      color: colors.textSecondary,
    },
    progressBarContainer: {
      width: '80%',
      height: 10,
      backgroundColor: colors.progressBarBackground,
      borderRadius: 5,
      overflow: 'hidden', // Ensures the gradient is clipped within the container
      marginBottom: spacing.md,
    },
    button: {
      backgroundColor: colors.secondary,
    },
    buttonText: {
      color: colors.buttonText,
    },
    animatedContainer: { height: '100%' },
  });

export default createStyles;
