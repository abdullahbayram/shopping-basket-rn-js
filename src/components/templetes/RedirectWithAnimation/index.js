import React, { useCallback, useEffect, useRef, useState } from 'react';
import { View, Animated, Text, BackHandler } from 'react-native';
import { useNavigation } from '@react-navigation/native';
import PropTypes from 'prop-types';
import Button from '../../atoms/Button';
import styles from './RedirectWithAnimation.style';

const RedirectWithAnimation = ({ message, duration = 5000, redirectTo = 'ProductList' }) => {
  const navigation = useNavigation();
  const [seconds, setSeconds] = useState(Math.max(duration / 1000, 1));
  const progress = useRef(new Animated.Value(0)).current;

  const resetNavigation = useCallback(() => {
    navigation.reset({
      index: 0,
      routes: [{ name: redirectTo }],
    });
  }, [navigation, redirectTo]);

  useEffect(() => {
    const backHandler = BackHandler.addEventListener('hardwareBackPress', () => true);

    Animated.timing(progress, {
      toValue: 1,
      duration,
      useNativeDriver: false,
    }).start();

    const interval = setInterval(() => {
      setSeconds((prev) => {
        if (prev <= 1) {
          clearInterval(interval);
        }
        return Math.max(prev - 1, 0);
      });
    }, 1000);

    const timer = setTimeout(() => {
      resetNavigation();
    }, duration);

    return () => {
      backHandler.remove();
      clearInterval(interval);
      clearTimeout(timer);
    };
  }, [navigation, progress, duration, resetNavigation]);

  const widthInterpolated = progress.interpolate({
    inputRange: [0, 1],
    outputRange: ['0%', '100%'],
  });

  return (
    <View style={styles.container}>
      <View style={styles.textContainer}>
        <Text style={styles.title}>{message}</Text>
        <Text style={styles.message}>
          Redirecting to product list in {seconds} second{seconds > 1 ? 's' : ''}...
        </Text>
      </View>
      <View style={styles.progressBarContainer}>
        <Animated.View style={[styles.progressBar, { width: widthInterpolated }]} />
      </View>
      <Button mode="contained" onPress={resetNavigation}>
        Go to Products
      </Button>
    </View>
  );
};

RedirectWithAnimation.propTypes = {
  message: PropTypes.string.isRequired,
  duration: PropTypes.number,
  redirectTo: PropTypes.string,
};

export default RedirectWithAnimation;
