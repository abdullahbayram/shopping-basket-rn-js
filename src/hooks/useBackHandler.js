import { useEffect } from 'react';
import { BackHandler } from 'react-native';

const useBackHandler = (callback) => {
  useEffect(() => {
    const backHandler = BackHandler.addEventListener('hardwareBackPress', callback);

    return () => {
      backHandler.remove();
    };
  }, [callback]);
};

export default useBackHandler;
