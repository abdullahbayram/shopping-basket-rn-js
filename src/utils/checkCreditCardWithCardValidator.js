const valid = require('card-validator');

const checkCreditCardWithCardValidator = (creditCardNumber) => {
  const numberValidation = valid.number(creditCardNumber);
  console.log(numberValidation, 'numberValidation');
  return numberValidation.isValid;
};

export default checkCreditCardWithCardValidator;
