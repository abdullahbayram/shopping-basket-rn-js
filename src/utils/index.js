export { default as fetchWithTimeout } from './fetchWithTimeout';
export { default as showToast } from './showToast';
export { default as paymentUtils } from './paymentUtils';
