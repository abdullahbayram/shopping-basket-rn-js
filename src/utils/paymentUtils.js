import { strings } from '../constants';

export default {
  filterNumericInput: (text) => text.replace(/[^0-9]/g, ''),
  formatExpirationDate: (value) => {
    const sanitized = value.replace(/[^0-9]/g, '');
    if (sanitized.length <= 2) return sanitized;
    return `${sanitized.slice(0, 2)}/${sanitized.slice(2, 4)}`;
  },
  getIcon: (field, errors, isCreditCardValid) => {
    console.log(errors, 'errors');
    const { icons } = strings; // Extract icons from strings
    let isValid = !errors[field]; // Check if the field is valid

    if (field === 'creditCardNumber') {
      isValid = isCreditCardValid;
    }
    return isValid ? icons[`${field}Check`] : icons[field]; // Dynamically map field to valid/invalid icons
  },
};
