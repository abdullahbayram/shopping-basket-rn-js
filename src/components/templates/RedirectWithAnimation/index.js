import React, { useCallback, useEffect, useRef, useState } from 'react';
import { View, Animated, StyleSheet } from 'react-native';
import { useNavigation } from '@react-navigation/native';
import PropTypes from 'prop-types';
import { useTheme } from 'react-native-paper';
import { spacing } from '@constants/theme';
import { strings } from '@constants';
import { Button, Text, LinearGradient } from '../../atoms';
import createStyles from './RedirectWithAnimation.style';
import useBackHandler from '../../../hooks/useBackHandler/useBackHandler';

const ANIMATION_DURATION = 1500;
const REDIRECT_DURATION = 5000;

const RedirectWithAnimation = ({ message, duration = REDIRECT_DURATION, redirectTo = 'ProductList' }) => {
  const { colors } = useTheme();
  const styles = createStyles(colors, spacing);
  const navigation = useNavigation();
  const [seconds, setSeconds] = useState(Math.max(duration / 1000, 1));
  const progress = useRef(new Animated.Value(0)).current;
  const shimmerPosition = useRef(new Animated.Value(-1)).current; // For the shimmer effect

  useBackHandler(() => true);

  const resetNavigation = useCallback(() => {
    navigation.reset({
      index: 0,
      routes: [{ name: redirectTo }],
    });
  }, [navigation, redirectTo]);

  useEffect(() => {
    Animated.timing(progress, {
      toValue: 1,
      duration,
      useNativeDriver: false,
    }).start();

    const shimmerAnimation = Animated.loop(
      Animated.timing(shimmerPosition, {
        toValue: 1,
        duration: ANIMATION_DURATION,
        useNativeDriver: false,
      }),
    );
    shimmerAnimation.start();

    const interval = setInterval(() => {
      setSeconds((prev) => Math.max(prev - 1, 0));
    }, 1000);

    const timer = setTimeout(() => {
      resetNavigation();
    }, duration);

    return () => {
      shimmerAnimation.stop();
      clearInterval(interval);
      clearTimeout(timer);
    };
  }, [navigation, progress, duration, resetNavigation, shimmerPosition]);

  const widthInterpolated = progress.interpolate({
    inputRange: [0, 1],
    outputRange: ['0%', '100%'],
  });

  const shimmerTranslate = shimmerPosition.interpolate({
    inputRange: [-1, 1],
    outputRange: [-100, 100],
  });

  return (
    <View style={styles.container}>
      <View style={styles.textContainer}>
        <Text variant="titleLarge" style={styles.title}>
          {message}
        </Text>
        <Text style={styles.message}>
          Redirecting to product list in {seconds} second{seconds > 1 ? 's' : ''}...
        </Text>
      </View>
      <View style={styles.progressBarContainer}>
        <Animated.View style={[styles.animatedContainer, { width: widthInterpolated }]}>
          <LinearGradient colors={[colors.linearLeft, colors.linearRight]} end={{ x: 1, y: 0 }} />
          <Animated.View
            style={{
              ...StyleSheet.absoluteFillObject,
              transform: [{ translateX: shimmerTranslate }],
            }}
          >
            <LinearGradient
              colors={['rgba(255, 255, 255, 0)', 'rgba(255, 255, 255, 0.3)', 'rgba(255, 255, 255, 0)']}
              end={{ x: 1, y: 0 }}
            />
          </Animated.View>
        </Animated.View>
      </View>
      <Button mode="contained" style={styles.button} textStyle={styles.buttonText} onPress={resetNavigation}>
        {strings.buttons.gotoProducts}
      </Button>
    </View>
  );
};

RedirectWithAnimation.propTypes = {
  message: PropTypes.string.isRequired,
  duration: PropTypes.number,
  redirectTo: PropTypes.string,
};

export default RedirectWithAnimation;
