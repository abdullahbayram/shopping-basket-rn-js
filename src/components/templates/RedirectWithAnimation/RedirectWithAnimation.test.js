import React from 'react';
import { screen, act, fireEvent } from '@testing-library/react-native';
import { strings } from '@constants';
import { renderInThemeProvider } from '../../../../__tests__/utils/renderInThemeProvider';
import RedirectWithAnimation from '.';
import useBackHandler from '../../../hooks/useBackHandler';

jest.mock('../../../hooks/useBackHandler', () => jest.fn());

jest.useFakeTimers();

const mockNavigation = {
  reset: jest.fn(),
};

jest.mock('@react-navigation/native', () => ({
  useNavigation: () => mockNavigation,
}));

describe('<RedirectWithAnimation />', () => {
  const message = 'Test message';

  beforeEach(() => {
    jest.clearAllMocks();
  });

  it('renders correctly and matches snapshot', () => {
    const { toJSON } = renderInThemeProvider(<RedirectWithAnimation message={message} />);

    expect(toJSON()).toMatchSnapshot();
  });

  it('renders the message and countdown', () => {
    renderInThemeProvider(<RedirectWithAnimation message={message} duration={3000} />);

    expect(screen.getByText(message)).toBeTruthy();
    expect(screen.getByText('Redirecting to product list in 3 seconds...')).toBeTruthy();
  });

  it('updates countdown every second', () => {
    renderInThemeProvider(<RedirectWithAnimation message={message} duration={3000} />);

    expect(screen.getByText('Redirecting to product list in 3 seconds...')).toBeTruthy();

    act(() => {
      jest.advanceTimersByTime(1000);
    });
    expect(screen.getByText('Redirecting to product list in 2 seconds...')).toBeTruthy();

    act(() => {
      jest.advanceTimersByTime(1000);
    });
    expect(screen.getByText('Redirecting to product list in 1 second...')).toBeTruthy();
  });

  it('navigates after duration', () => {
    renderInThemeProvider(<RedirectWithAnimation message={message} duration={3000} />);

    act(() => {
      jest.advanceTimersByTime(3000);
    });

    expect(mockNavigation.reset).toHaveBeenCalledWith({
      index: 0,
      routes: [{ name: 'ProductList' }],
    });
  });

  it('handles button press to navigate immediately', () => {
    renderInThemeProvider(<RedirectWithAnimation message={message} />);

    const button = screen.getByText(strings.buttons.gotoProducts);
    expect(button).toBeTruthy();

    fireEvent.press(button);

    expect(mockNavigation.reset).toHaveBeenCalledWith({
      index: 0,
      routes: [{ name: 'ProductList' }],
    });
  });

  it('disables back button functionality', () => {
    renderInThemeProvider(<RedirectWithAnimation message={message} />);

    // Simulate the back handler callback
    const backHandlerCallback = useBackHandler.mock.calls[0][0];
    const result = backHandlerCallback();

    // Verify that the back handler prevents navigation
    expect(result).toBe(true);
  });
});
