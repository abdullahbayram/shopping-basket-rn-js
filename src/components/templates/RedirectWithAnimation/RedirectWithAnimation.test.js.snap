// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<RedirectWithAnimation /> renders correctly and matches snapshot 1`] = `
<View
  style={
    {
      "alignItems": "center",
      "backgroundColor": "#FFF",
      "flex": 1,
      "justifyContent": "center",
    }
  }
>
  <View
    style={
      {
        "alignItems": "center",
        "marginBottom": 24,
      }
    }
  >
    <Text
      style={
        [
          {
            "textAlign": "left",
          },
          {
            "color": "rgba(28, 27, 31, 1)",
            "writingDirection": "ltr",
          },
          [
            {
              "fontFamily": "System",
              "fontSize": 22,
              "fontWeight": "400",
              "letterSpacing": 0,
              "lineHeight": 28,
            },
            [
              {
                "color": "#34300F",
              },
              {
                "color": "#004f6b",
                "fontSize": 24,
                "fontWeight": "700",
                "marginBottom": 7,
                "textAlign": "center",
              },
            ],
          ],
        ]
      }
      testID="text"
    >
      Test message
    </Text>
    <Text
      style={
        [
          {
            "textAlign": "left",
          },
          {
            "color": "rgba(28, 27, 31, 1)",
            "writingDirection": "ltr",
          },
          [
            {
              "fontFamily": "System",
              "fontSize": 14,
              "fontWeight": "400",
              "letterSpacing": 0.25,
              "lineHeight": 20,
            },
            [
              {
                "color": "#34300F",
              },
              {
                "color": "#666",
                "fontSize": 16,
              },
            ],
          ],
        ]
      }
      testID="text"
    >
      Redirecting to product list in 
      5
       second
      s
      ...
    </Text>
  </View>
  <View
    style={
      {
        "backgroundColor": "#e0e0e0",
        "borderRadius": 5,
        "height": 10,
        "marginBottom": 16,
        "overflow": "hidden",
        "width": "80%",
      }
    }
  >
    <View
      collapsable={false}
      style={
        {
          "height": "100%",
          "width": "0%",
        }
      }
    >
      <ViewManagerAdapter_ExpoLinearGradient
        colors={
          [
            4285515996,
            4287354055,
          ]
        }
        endPoint={
          [
            1,
            0,
          ]
        }
        startPoint={
          [
            0,
            0,
          ]
        }
        style={
          [
            {
              "flex": 1,
            },
            {},
          ]
        }
        testID="linear-gradient"
      />
      <View
        collapsable={false}
        style={
          {
            "bottom": 0,
            "left": 0,
            "position": "absolute",
            "right": 0,
            "top": 0,
            "transform": [
              {
                "translateX": -100,
              },
            ],
          }
        }
      >
        <ViewManagerAdapter_ExpoLinearGradient
          colors={
            [
              16777215,
              1308622847,
              16777215,
            ]
          }
          endPoint={
            [
              1,
              0,
            ]
          }
          startPoint={
            [
              0,
              0,
            ]
          }
          style={
            [
              {
                "flex": 1,
              },
              {},
            ]
          }
          testID="linear-gradient"
        />
      </View>
    </View>
  </View>
  <View
    collapsable={false}
    style={
      {
        "backgroundColor": "#FED034",
        "borderRadius": 10,
        "shadowColor": "#000",
        "shadowOffset": {
          "height": 0,
          "width": 0,
        },
        "shadowOpacity": 0,
        "shadowRadius": 0,
      }
    }
    testID="button-container-outer-layer"
  >
    <View
      collapsable={false}
      style={
        {
          "backgroundColor": "#FED034",
          "borderColor": "transparent",
          "borderRadius": 10,
          "borderStyle": "solid",
          "borderWidth": 0,
          "flex": undefined,
          "minWidth": 64,
          "shadowColor": "#000",
          "shadowOffset": {
            "height": 0,
            "width": 0,
          },
          "shadowOpacity": 0,
          "shadowRadius": 0,
        }
      }
      testID="button-container"
    >
      <View
        accessibilityRole="button"
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": false,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          [
            {
              "overflow": "hidden",
            },
            {
              "borderRadius": 10,
            },
          ]
        }
        testID="button"
      >
        <View
          style={
            [
              {
                "alignItems": "center",
                "flexDirection": "row",
                "justifyContent": "center",
              },
              undefined,
            ]
          }
        >
          <Text
            numberOfLines={1}
            selectable={false}
            style={
              [
                {
                  "textAlign": "left",
                },
                {
                  "color": "rgba(28, 27, 31, 1)",
                  "writingDirection": "ltr",
                },
                [
                  {
                    "fontFamily": "System",
                    "fontSize": 14,
                    "fontWeight": "500",
                    "letterSpacing": 0.1,
                    "lineHeight": 20,
                  },
                  [
                    {
                      "marginHorizontal": 16,
                      "marginVertical": 9,
                      "textAlign": "center",
                    },
                    false,
                    {
                      "marginHorizontal": 24,
                      "marginVertical": 10,
                    },
                    undefined,
                    false,
                    {
                      "color": "#34300F",
                      "fontFamily": "System",
                      "fontSize": 14,
                      "fontWeight": "500",
                      "letterSpacing": 0.1,
                      "lineHeight": 20,
                    },
                    undefined,
                  ],
                ],
              ]
            }
            testID="button-text"
          >
            Go to Products
          </Text>
        </View>
      </View>
    </View>
  </View>
</View>
`;
